pkg/runtime/serializer/codec_factory.go

```diff
@@ -17,36 +17,36 @@ limitations under the License.
 package serializer
 
 import (
+	"io/ioutil"
+
 	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/runtime/serializer/json"
 	"k8s.io/kubernetes/pkg/runtime/serializer/recognizer"
+	"k8s.io/kubernetes/pkg/runtime/serializer/streaming"
 	"k8s.io/kubernetes/pkg/runtime/serializer/versioning"
 )
 
+// serializerExtensions are for serializers that are conditionally compiled in
+var serializerExtensions = []func(*runtime.Scheme) (serializerType, bool){}
+
 type serializerType struct {
 	AcceptContentTypes []string
 	ContentType        string
 	FileExtensions     []string
-	Serializer         runtime.Serializer
-	PrettySerializer   runtime.Serializer
-}
 
-// NewCodecFactory provides methods for retrieving serializers for the supported wire formats
-// and conversion wrappers to define preferred internal and external versions. In the future,
-// as the internal version is used less, callers may instead use a defaulting serializer and
-// only convert objects which are shared internally (Status, common API machinery).
-// TODO: allow other codecs to be compiled in?
-// TODO: accept a scheme interface
-func NewCodecFactory(scheme *runtime.Scheme) CodecFactory {
-	return newCodecFactory(scheme, json.DefaultMetaFactory)
+	Serializer       runtime.Serializer
+	PrettySerializer runtime.Serializer
+	// If set, will be used for streaming codecs. This serializer is assumed to not need explicit
+	// typing via a wrapper.
+	ImplicitSerializer runtime.Serializer
+
+	Specialize func(map[string]string) (runtime.Serializer, bool)
```

Comment:
Names probably aren't very good, this takes the extra flags provided on the
mime type and passes them to a function which can do more specific checks.
So if we had a content-type
"application/vnd.kubernetes.protobuf;type=Event" the map {"type": "Event"}
is passed to specialize.  It's used today for "pretty".  Will add a comment.

On Fri, Feb 26, 2016 at 7:44 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/runtime/serializer/codec_factory.go
> <https://github.com/kubernetes/kubernetes/pull/20377#discussion_r54243769>
> :
>
> >
> > -// NewCodecFactory provides methods for retrieving serializers for the supported wire formats
> > -// and conversion wrappers to define preferred internal and external versions. In the future,
> > -// as the internal version is used less, callers may instead use a defaulting serializer and
> > -// only convert objects which are shared internally (Status, common API machinery).
> > -// TODO: allow other codecs to be compiled in?
> > -// TODO: accept a scheme interface
> > -func NewCodecFactory(scheme *runtime.Scheme) CodecFactory {
> > -	return newCodecFactory(scheme, json.DefaultMetaFactory)
> > +	Serializer       runtime.Serializer
> > +	PrettySerializer runtime.Serializer
> > +	// If set, will be used for streaming codecs. This serializer is assumed to not need explicit
> > +	// typing via a wrapper.
> > +	ImplicitSerializer runtime.Serializer
> > +
> > +	Specialize func(map[string]string) (runtime.Serializer, bool)
>
> What's that? Can you please add a comment?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20377/files#r54243769>.
>